pkg/apps/packagekit.js

```diff
@@ -0,0 +1,185 @@
+                        data.waiting = true;
+                    if (tr.Percentage !== undefined && tr.Percentage !== 101)
+                        data.percentage = tr.Percentage;
+                    if (tr.AllowCancel)
+                        data.cancel = cancel;
+
+                    progress_cb(data);
+                }
+            }
+
+            changed();
+            $(tr).on("changed", changed);
+            window.setTimeout(function () {
+                allow_wait_status = true;
+                changed();
+            }, 1000);
+
+            $(tr).on("ErrorCode", function (event, code, details) {
+                defer.reject(details, cancelled? "cancelled" : code);
+            });
+            $(tr).on("Package", function (event, info, package_id, summary) {
+                if (package_cb && defer.promise().state() == "pending")
+                    package_cb(info, package_id, summary);
+            });
+            $(tr).on("Finished", function (event, exit, runtime) {
+                defer.resolve(exit);
+            });
+            tr.call(method, args).fail(function (error) {
+                console.log("Error", error);
+                defer.reject(error);
+            });
+        }).
+        fail(function (error) {
+            defer.reject(error);
+        });
+
+    return defer.promise();
+}
+
+function progress_reporter(base, range, callback) {
+    if (callback)
+        return function (data) {
+            if (data.percentage >= 0)
+                data.percentage = base + data.percentage/100*range;
+            callback(data);
+        };
+}
+
+function resolve(method, filter, name, progress_cb) {
+    var defer = cockpit.defer();
+    var ids = [ ];
+
+    function gather_package_cb(info, package_id) {
+        ids.push(package_id);
+    }
+
+    transaction(method, [ filter, [ name ] ], progress_cb, gather_package_cb).
+        done(function () {
+            if (ids.length === 0)
+                defer.reject("Can't resolve package", "not-found");
+            else
+                defer.resolve(ids[0]);
+        }).
+        fail(function (error) {
+            defer.reject(error);
+        });
+
+    return defer.promise();
+}
+
+function reload_bridge_packages() {
+    return cockpit.dbus(null, { bus: "internal" }).call("/packages", "cockpit.Packages", "Reload", [ ]);
+}
+
+function install(name, progress_cb) {
+    return resolve("Resolve", PK_FILTER_ARCH | PK_FILTER_NOT_SOURCE | PK_FILTER_NEWEST, name,
+                   progress_reporter(0, 10, progress_cb)).
+        then(function (pkgid) {
+            return transaction("InstallPackages", [ 0, [ pkgid ] ], progress_reporter(10, 90, progress_cb)).
+                then(reload_bridge_packages);
+        });
+}
+
+function remove(name, progress_cb) {
+    return resolve("SearchFiles", PK_FILTER_INSTALLED, name, progress_reporter(0, 10, progress_cb)).
```

Comment:
The package name comes from the repository meta data (if we treat the collection AppStream as part of the repository meta data as we should I'd say), and we might not have this meta data for all installed packages.  The package might not have been installed from a repository in the first place, or the repository might have been disabled, or it might have changed too much.